src/puppetlabs/puppetdb/command.clj

```diff
@@ -75,21 +74,130 @@
             [clojure.set :as set]
             [clojure.core.async :as async]
             [metrics.timers :refer [timer time!]]
-            [metrics.counters :refer [inc!]]
+            [metrics.counters :refer [inc! dec! counter]]
+            [metrics.meters :refer [meter mark!]]
+            [metrics.histograms :refer [histogram update!]]
             [puppetlabs.puppetdb.metrics.core :as metrics]
             [puppetlabs.puppetdb.queue :as queue]
             [clj-time.coerce :as tcoerce]
             [puppetlabs.puppetdb.amq-migration :as mig]
-            [puppetlabs.puppetdb.command.dlo :as dlo]))
+            [puppetlabs.puppetdb.command.dlo :as dlo]
+            [overtone.at-at :refer [mk-pool stop-and-reset-pool! after]]
+            [puppetlabs.puppetdb.threadpool :as gtp]
+            [puppetlabs.puppetdb.utils.metrics :as mutils]))
+
+;; ## Performance counters
+
+;; For each command (and globally), add command-processing
+;; metrics. The hierarchy of command-processing metrics has the
+;; following form:
+;;
+;;     {"global"
+;;        {:seen <meter>
+;;         :processed <meter>
+;;         :fatal <meter>
+;;         :retried <meter>
+;;         :discarded <meter>
+;;         :processing-time <timer>
+;;         :retry-counts <histogram>
+;;        }
+;;      "command name"
+;;        {<version number>
+;;           {:seen <meter>
+;;            :processed <meter>
+;;            :fatal <meter>
+;;            :retried <meter>
+;;            :discarded <meter>
+;;            :processing-time <timer>
+;;            :retry-counts <histogram>
+;;           }
+;;        }
+;;     }
+;;
+;; The `"global"` hierarchy contains metrics aggregated across all
+;; commands.
+;;
+;; * `:seen`: the number of commands (valid or not) encountered
+;;
+;; * `:processeed`: the number of commands successfully processed
+;;
+;; * `:fatal`: the number of fatal errors
+;;
+;; * `:retried`: the number of commands re-queued for retry
+;;
+;; * `:discarded`: the number of commands discarded for exceeding the
+;;   maximum allowable retry count
+;;
+;; * `:processing-time`: how long it takes to process a command
+;;
+;; * `:retry-counts`: histogram containing the number of times
+;;   messages have been retried prior to suceeding
+;;
```

Comment:
I didn't change any of this, just moved it over, but I'll add it, maybe reformat some of those comments to not take up so much space